### PR TITLE
Fix 525 type collision modulo case

### LINKED ISSUE #525: Identical classes modulo case name in pulumi-azure-native v1.63.0
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

It seems that we have two identical enums that differ only in the case of the class name in the schema.

This runs into issues for Mac OS X users, since Mac OS FS is not case-sensitive, running codegen on Mac OS X confuses the two files. One of these enums clobbers the other one.

```
anton@Antons-MacBook-Pro> jq '.types["azure-native:network:IpAllocationMethod"]' ~/schemas/pulumi-azure-native-v1.63.0.json                                        ~/pulumi-java
{
  "description": "Private IP address allocation method.",
  "type": "string",
  "enum": [
    {
      "value": "Static"
    },
    {
      "value": "Dynamic"
    }
  ]
}
anton@Antons-MacBook-Pro> jq '.types["azure-native:network:IPAllocationMethod"]' ~/schemas/pulumi-azure-native-v1.63.0.json                                        ~/pulumi-java
{
  "description": "The private IP address allocation method.",
  "type": "string",
  "enum": [
    {
      "value": "Static"
    },
    {
      "value": "Dynamic"
    }
  ]
}
```

There is a chance it is a schema issue that need to be fixed in pulumi-azure-native to unify the two named types.

### PR DESCRIPTION
**Pull request by t0yv0**
# Description

Fixes #525 

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/cmd/pulumi-java-gen/dedup.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+// Detects cases when identical types have similar names modulo case
+// such as `azure-native:network:IpAllocationMethod` vs
+// `azure-native:network:IPAllocationMethod`, deterministically picks
+// one of these names, and rewrites the schema as if there was only
+// one such type.
+func dedupTypes(spec *pschema.PackageSpec) (*pschema.PackageSpec, error) {
+	normalizedTokens := map[string]string{}
+	for typeToken := range spec.Types {
```
Comment (iwahbe) [inline pkg/cmd/pulumi-java-gen/dedup.go:22]: We do this for type names but not for resource or function tokens?

**Review by iwahbe — COMMENTED** [review]
I think this works, but I'm concerned with the mechanism:
 It seems problematic to rewrite the raw schema before the schema is loaded, especially since we won't always control how the schema is loaded (TF bridge). I'm also concerned that it will prevent java from using our standard package loading infrastructure (`schema.Loader`). Is this programgen compatible?

**Comment by t0yv0** [comment]
You're correct that programgen would continue emitting mis-spelled code for the affected type. I am inclined to ignore this for the moment. I still think this is the appropriate fix since the blast radius of the bug is exactly N=1 type in azure-native:

- no functions
- no resources
- Java provider generation do not use tfbridge, it uses the schema as input, where tfbrdige was used to make the schema
- https://github.com/pulumi/pulumi-azure-native/issues/1692 would be preferable to fix upstream, then this fix can be dropped

**Review by iwahbe — APPROVED** [review]
I buy that. Once azure-native is fixed we can rip this out and hopefully make these collisions illegal at the schema level. 
That may be a 4.0 level change.

**Pull request merged**